8.0+ currently breaks compatibility with unconventional methods to load mixins for mods

This is precisely why MixinBooter was introduced in the first place. This issue replaces #30, #32 and #33 at the present time.

Currently effected mods:

- [x] Just Enough IDs
- [x] Roughly Enough IDs
- [x] Integrated Proxy
- [x] DJ2 Addons
- [x] ThaumicFixes
- [x] ErebusFix
- [ ] RLMixins
- [ ] VanillaFix
- [ ] Grimoire
- [ ] MUP
- [ ] MakeZoomZoom
- [ ] DupeFix Project

so far it seems to work on client but my server is having a critical crash error:

[20:13:02] [main/TRACE] [mixin]: Added class metadata for java/lang/Runnable to metadata cache
[20:13:02] [main/ERROR] [LaunchWrapper]: Unable to launch
java.lang.ClassNotFoundException: net.minecraft.server.MinecraftServer
	at net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:191) ~[launchwrapper-1.12.jar:?]
	at java.lang.ClassLoader.loadClass(Unknown Source) ~[?:1.8.0_371]
	at java.lang.ClassLoader.loadClass(Unknown Source) ~[?:1.8.0_371]
	at java.lang.Class.forName0(Native Method) ~[?:1.8.0_371]
	at java.lang.Class.forName(Unknown Source) ~[?:1.8.0_371]
	at net.minecraft.launchwrapper.Launch.launch(Launch.java:131) ~[launchwrapper-1.12.jar:?]
	at net.minecraft.launchwrapper.Launch.main(Launch.java:28) ~[launchwrapper-1.12.jar:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_371]
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source) ~[?:1.8.0_371]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[?:1.8.0_371]
	at java.lang.reflect.Method.invoke(Unknown Source) ~[?:1.8.0_371]
	at net.minecraftforge.fml.relauncher.ServerLaunchWrapper.run(ServerLaunchWrapper.java:148) ~[Magma-1.12.2-165ec152.jar:1.12.2-d174ef5f]
	at net.minecraftforge.fml.relauncher.ServerLaunchWrapper.main(ServerLaunchWrapper.java:100) ~[Magma-1.12.2-165ec152.jar:1.12.2-d174ef5f]
Caused by: org.spongepowered.asm.mixin.transformer.throwables.MixinTransformerError: An unexpected critical error was encountered
	at org.spongepowered.asm.mixin.transformer.MixinProcessor.applyMixins(MixinProcessor.java:392) ~[!mixinbooter-8.2.jar:?]
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClass(MixinTransformer.java:234) ~[!mixinbooter-8.2.jar:?]
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClassBytes(MixinTransformer.java:202) ~[!mixinbooter-8.2.jar:?]
	at org.spongepowered.asm.mixin.transformer.Proxy.transform(Proxy.java:72) ~[!mixinbooter-8.2.jar:?]
	at net.minecraft.launchwrapper.LaunchClassLoader.runTransformers(LaunchClassLoader.java:279) ~[launchwrapper-1.12.jar:?]
	at net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:176) ~[launchwrapper-1.12.jar:?]
	... 12 more
Caused by: org.spongepowered.asm.mixin.injection.throwables.InjectionError: Critical injection failure: Redirector setPriorityAndStart(Ljava/lang/Thread;)V in mixins.priorities.json:MinecraftServerMixin from mod censoredasm5_13 failed injection check, (0/1) succeeded. Scanned 1 target(s). Using refmap mixins.loliasm.refmap.json
	at org.spongepowered.asm.mixin.injection.struct.InjectionInfo.postInject(InjectionInfo.java:468) ~[!mixinbooter-8.2.jar:?]
	at org.spongepowered.asm.mixin.transformer.MixinTargetContext.applyInjections(MixinTargetContext.java:1384) ~[!mixinbooter-8.2.jar:?]
	at org.spongepowered.asm.mixin.transformer.MixinApplicatorStandard.applyInjections(MixinApplicatorStandard.java:1062) ~[!mixinbooter-8.2.jar:?]
	at org.spongepowered.asm.mixin.transformer.MixinApplicatorStandard.applyMixin(MixinApplicatorStandard.java:402) ~[!mixinbooter-8.2.jar:?]
	at org.spongepowered.asm.mixin.transformer.MixinApplicatorStandard.apply(MixinApplicatorStandard.java:327) ~[!mixinbooter-8.2.jar:?]
	at org.spongepowered.asm.mixin.transformer.TargetClassContext.apply(TargetClassContext.java:421) ~[!mixinbooter-8.2.jar:?]
	at org.spongepowered.asm.mixin.transformer.TargetClassContext.applyMixins(TargetClassContext.java:403) ~[!mixinbooter-8.2.jar:?]
	at org.spongepowered.asm.mixin.transformer.MixinProcessor.applyMixins(MixinProcessor.java:363) ~[!mixinbooter-8.2.jar:?]
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClass(MixinTransformer.java:234) ~[!mixinbooter-8.2.jar:?]
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClassBytes(MixinTransformer.java:202) ~[!mixinbooter-8.2.jar:?]
	at org.spongepowered.asm.mixin.transformer.Proxy.transform(Proxy.java:72) ~[!mixinbooter-8.2.jar:?]
	at net.minecraft.launchwrapper.LaunchClassLoader.runTransformers(LaunchClassLoader.java:279) ~[launchwrapper-1.12.jar:?]
	at net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:176) ~[launchwrapper-1.12.jar:?]
	... 12 more
[20:13:02] [main/INFO] [STDERR]: [net.minecraftforge.fml.relauncher.ServerLaunchWrapper:main:100]: java.lang.reflect.InvocationTargetException
[20:13:02] [main/INFO] [STDERR]: [net.minecraftforge.fml.relauncher.ServerLaunchWrapper:main:100]: 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[20:13:02] [main/INFO] [STDERR]: [net.minecraftforge.fml.relauncher.ServerLaunchWrapper:main:100]: 	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
[20:13:02] [main/INFO] [STDERR]: [net.minecraftforge.fml.relauncher.ServerLaunchWrapper:main:100]: 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
[20:13:02] [main/INFO] [STDERR]: [net.minecraftforge.fml.relauncher.ServerLaunchWrapper:main:100]: 	at java.lang.reflect.Method.invoke(Unknown Source)
[20:13:02] [main/INFO] [STDERR]: [net.minecraftforge.fml.relauncher.ServerLaunchWrapper:main:100]: 	at net.minecraftforge.fml.relauncher.ServerLaunchWrapper.run(ServerLaunchWrapper.java:148)
[20:13:02] [main/INFO] [STDERR]: [net.minecraftforge.fml.relauncher.ServerLaunchWrapper:main:100]: 	at net.minecraftforge.fml.relauncher.ServerLaunchWrapper.main(ServerLaunchWrapper.java:100)
[20:13:02] [main/INFO] [STDERR]: [net.minecraftforge.fml.relauncher.ServerLaunchWrapper:run:151]: Caused by: net.minecraftforge.fml.relauncher.FMLSecurityManager$ExitTrappedException
[20:13:02] [main/INFO] [STDERR]: [net.minecraftforge.fml.relauncher.ServerLaunchWrapper:run:151]: 	at net.minecraftforge.fml.relauncher.FMLSecurityManager.checkPermission(FMLSecurityManager.java:49)
[20:13:02] [main/INFO] [STDERR]: [net.minecraftforge.fml.relauncher.ServerLaunchWrapper:run:151]: 	at java.lang.SecurityManager.checkExit(Unknown Source)
[20:13:02] [main/INFO] [STDERR]: [net.minecraftforge.fml.relauncher.ServerLaunchWrapper:run:151]: 	at java.lang.Runtime.exit(Unknown Source)
[20:13:02] [main/INFO] [STDERR]: [net.minecraftforge.fml.relauncher.ServerLaunchWrapper:run:151]: 	at java.lang.System.exit(Unknown Source)
[20:13:02] [main/INFO] [STDERR]: [net.minecraftforge.fml.relauncher.ServerLaunchWrapper:run:151]: 	at net.minecraft.launchwrapper.Launch.launch(Launch.java:138)
[20:13:02] [main/INFO] [STDERR]: [net.minecraftforge.fml.relauncher.ServerLaunchWrapper:run:151]: 	at net.minecraft.launchwrapper.Launch.main(Launch.java:28)
[20:13:02] [main/INFO] [STDERR]: [net.minecraftforge.fml.relauncher.ServerLaunchWrapper:run:151]: 	... 6 more

--- a/src/launchwrapper/java/org/spongepowered/asm/mixin/transformer/Proxy.java
+++ b/src/launchwrapper/java/org/spongepowered/asm/mixin/transformer/Proxy.java
@@ -24,13 +24,15 @@
  */
 package org.spongepowered.asm.mixin.transformer;
 
-import java.util.ArrayList;
-import java.util.List;
-
+import net.minecraft.launchwrapper.IClassTransformer;
+import org.spongepowered.asm.logging.ILogger;
+import org.spongepowered.asm.mixin.MixinEnvironment;
+import org.spongepowered.asm.mixin.extensibility.IMixinProcessor;
 import org.spongepowered.asm.service.ILegacyClassTransformer;
 import org.spongepowered.asm.service.MixinService;
 
-import net.minecraft.launchwrapper.IClassTransformer;
+import java.util.ArrayList;
+import java.util.List;
 
 /**
  * Proxy transformer for the mixin transformer. These transformers are used
@@ -93,5 +95,25 @@ public final class Proxy implements IClassTransformer, ILegacyClassTransformer {
         
         return basicClass;
     }
-
+    
+    /**
+     * Tells the Mixin subsystem to process any Mixin configs that were added since we last checked. Mixins applied this way still cannot transform classes that are already loaded.
+     * <p>A common use-case is to apply configs targeting classes that do not exist in the classloader before Mixin first initializes.</p>
+     */
+    public static void processLateConfigs() {
+        final IMixinProcessor processor = transformer.getProcessor();
+        final ILogger logger = MixinService.getService().getLogger("mixin");
+        if (processor.getPendingMixinConfigs().isEmpty()) {
+            logger.warn("Proxy::processLateConfigs was called with no pending configs.");
+            return;
+        }
+        if (!(processor instanceof MixinProcessor)) {
+            logger.error("Proxy.transformer.processor is not a MixinProcessor. Could not apply late configs.");
+            return;
+        }
+        // NOTE: If anyone has a better way of doing this, please change this to use it. I'm just making a jankless version of the reflection-based method we've been using for years.
+        final MixinProcessor mixinProcessor = (MixinProcessor) processor;
+        mixinProcessor.selectConfigs(MixinEnvironment.getCurrentEnvironment());
+        mixinProcessor.prepareConfigs(MixinEnvironment.getCurrentEnvironment(), mixinProcessor.extensions);
+    }
 }

--- a/src/main/java/org/spongepowered/asm/mixin/transformer/MixinProcessor.java
+++ b/src/main/java/org/spongepowered/asm/mixin/transformer/MixinProcessor.java
@@ -68,7 +68,7 @@ import org.spongepowered.asm.util.perf.Profiler.Section;
 /**
  * Heart of the Mixin pipeline 
  */
-class MixinProcessor implements IMixinProcessor {
+public class MixinProcessor implements IMixinProcessor {
 
     /**
      * Phase during which an error occurred, delegates to functionality in
@@ -595,6 +595,17 @@ class MixinProcessor implements IMixinProcessor {
         this.pendingConfigs.clear();
         
         return totalMixins;
+    }
+    
+    /**
+     * @deprecated Compatibility with common brittle Reflection-based usages pre-0.8.
+     * @param environment Environment
+     * @return Total number of Mixins initialized
+     */
+    @Deprecated
+    private int prepareConfigs(MixinEnvironment environment) {
+        MixinProcessor.logger.warn("MixinTransformer::prepareConfigs(MixinEnvironment) is deprecated!");
+        return prepareConfigs(environment, this.extensions);
     }
 
     private void handleMixinPrepareError(MixinConfig config, InvalidMixinException ex, MixinEnvironment environment) throws MixinPrepareError {

--- a/src/main/java/org/spongepowered/asm/mixin/transformer/MixinProcessor.java
+++ b/src/main/java/org/spongepowered/asm/mixin/transformer/MixinProcessor.java
@@ -173,7 +173,7 @@ public class MixinProcessor implements IMixinProcessor {
     /**
      * Processor extensions
      */
-    private final Extensions extensions;
+    public final Extensions extensions;
     
     /**
      * Hot-Swap agent
@@ -501,7 +501,7 @@ public class MixinProcessor implements IMixinProcessor {
      * 
      * @param environment Environment to query
      */
-    private void selectConfigs(MixinEnvironment environment) {
+    public void selectConfigs(MixinEnvironment environment) {
         for (Iterator<Config> iter = Mixins.getConfigs().iterator(); iter.hasNext();) {
             Config handle = iter.next();
             try {
@@ -521,12 +521,12 @@ public class MixinProcessor implements IMixinProcessor {
     }
 
     /**
-     * Prepare mixin configs
+     * Prepares pending mixin configs
      * 
      * @param environment Environment
      * @return total number of mixins initialised
      */
-    private int prepareConfigs(MixinEnvironment environment, Extensions extensions) {
+    protected int prepareConfigs(MixinEnvironment environment, Extensions extensions) {
         int totalMixins = 0;
         
         final IHotSwap hotSwapper = this.hotSwapper;
@@ -598,13 +598,14 @@ public class MixinProcessor implements IMixinProcessor {
     }
     
     /**
-     * @deprecated Compatibility with common brittle Reflection-based usages pre-0.8.
+     * @deprecated Added for compatibility with unfortunately-common brittle Reflection-based usages pre-0.8.
+     * @see MixinProcessor#prepareConfigs(org.spongepowered.asm.mixin.MixinEnvironment, org.spongepowered.asm.mixin.transformer.ext.Extensions)
      * @param environment Environment
      * @return Total number of Mixins initialized
      */
     @Deprecated
     private int prepareConfigs(MixinEnvironment environment) {
-        MixinProcessor.logger.warn("MixinTransformer::prepareConfigs(MixinEnvironment) is deprecated!");
+        MixinProcessor.logger.warn("MixinProcessor::prepareConfigs(MixinEnvironment) is deprecated!");
         return prepareConfigs(environment, this.extensions);
     }
 

--- a/src/main/java/org/spongepowered/asm/mixin/transformer/MixinTransformer.java
+++ b/src/main/java/org/spongepowered/asm/mixin/transformer/MixinTransformer.java
@@ -43,7 +43,7 @@ import org.spongepowered.asm.util.asm.ASM;
 /**
  * Transformer which manages the mixin configuration and application process
  */
-final class MixinTransformer extends TreeTransformer implements IMixinTransformer {
+public final class MixinTransformer extends TreeTransformer implements IMixinTransformer {
     
     /**
      * Impl of mixin transformer factory

